### PR TITLE
Improve Makefile and code quality

### PR DESCRIPTION
**Pull request by lvwerra**
This PR improves Makefile commands (borrowing from evaluate/datasets) and adds a code quality CI. The last commit fixes the existing code quality issues.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -7,7 +7,27 @@ on:
...
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.7"
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install .[dev]
+      - name: Check quality
+        run: |
+          black --check --line-length 119 --target-version py36 tests trl
```
Comment (lewtun) [inline .github/workflows/tests.yml:25]: If I'm not mistaken, you can just run `make quality` here. this has the advantage of having a single source of truth for the quality logic

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -2,9 +2,10 @@
 
 ## How to get started
 
-Before anything else, please install the git hooks that run automatic scripts during each commit and merge to strip the notebooks of superfluous metadata (and avoid merge conflicts). After cloning the repository, run the following command inside it:
-```
-nbdev_install_git_hooks
+Before you start contributing make sure you installed all the dev tools:
+
+```bash
+pip install "trl[dev]"
```
Comment (lewtun) [inline CONTRIBUTING.md:8]: For contributors, I think you want to have an editable install:

```suggestion
pip install -e ".[dev]"
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -27,7 +28,21 @@ nbdev_install_git_hooks
...
+Then before submitting your PR make sure the code quality follows the standards. You can run the following command to format and test:
+
+```bash
+make style && make quality
+```
+
 ## Do you want to contribute to the documentation?
 
-* Docs are automatically created from the notebooks in the nbs folder.
+* Docs are are in the `docs/` folder and can be updated there.
```
Comment (lewtun) [inline CONTRIBUTING.md:47]:
```suggestion
* Docs are in the `docs/` folder and can be updated there.
```

File: Makefile [hunk 4]
```diff
@@ -1,33 +1,13 @@
...
-	nbdev_bump_version
-
-pypi: dist
-	twine upload --repository pypi dist/*
+	python -m pytest -n auto --dist=loadfile -s -v ./tests/
 
-dist: clean
-	python setup.py sdist bdist_wheel
+quality:
+	black --check --line-length 119 --target-version py36 tests trl
```
Comment (lewtun) [inline Makefile:7]: python 3.6 is dead and 3.7 will be eol in 6 months ([link](https://endoflife.date/python)). maybe good to already start with 3.8 as the minimum version?
```suggestion
	black --check --line-length 119 --target-version py38 tests trl
```

File: .github/workflows/tests.yml [hunk 5]
```diff
@@ -7,7 +7,27 @@ on:
...
 jobs:
+
+  check_code_quality:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python
+        uses: actions/setup-python@v4
+        with:
+          python-version: "3.7"
```
Comment (lewtun) [inline .github/workflows/tests.yml:18]: see my comment below about considering v3.8 as the minimum version since this one will be eol in 6 months

File: Makefile [hunk 6]
```diff
@@ -1,33 +1,13 @@
...
+	isort --check-only tests trl
+	flake8  tests trl
 
-clean:
-	rm -rf dist
\ No newline at end of file
+style:
+	black --line-length 119 --target-version py36 tests trl
```
Comment (lewtun) [inline Makefile:12]: same thing here if you agree with the above
```suggestion
	black --line-length 119 --target-version py38 tests trl
```

File: tests/test_modeling.py [hunk 7]
```diff
@@ -60,19 +61,25 @@ def test_from_save_transformers(self):
...
             for key in transformers_model.state_dict():
-                self.assertTrue(torch.allclose(transformers_model_from_save.state_dict()[key], transformers_model.state_dict()[key]))
+                self.assertTrue(
+                    torch.allclose(
+                        transformers_model_from_save.state_dict()[key], transformers_model.state_dict()[key]
+                    )
+                )
+
 
 class VHeadModelTester(BaseModelTester, unittest.TestCase):
```
Comment (lewtun) [inline tests/test_modeling.py:78]: not for this pr, but a small nit that being explicit with e.g. `ValueHeadModelTester` is generally best practice + enables newcomers to easier understand the codebase. 

the same nit applies to docstrings using `v-head` vs `value head` and filenames like `modeling_vhead.py` being better named as `modeling_value_head.py`

File: tests/test_modeling.py [hunk 8]
```diff
@@ -83,25 +90,25 @@ def test_vhead(self):
         for model_name in self.all_model_names:
             model = AutoModelForCausalLMWithValueHead.from_pretrained(model_name)
             self.assertTrue(hasattr(model, "v_head"))
-    
+
     def test_vhead_nb_classes(self):
```
Comment (lewtun) [inline tests/test_modeling.py:94]: would this be better?
```suggestion
    def test_vhead_shape(self):
```

**Review by lewtun — COMMENTED** [review]
Thanks for improving the quality of the codebase @lvwerra 🕺 ! I left some suggestions to bump to python 3.8 to avoid an annoying refactor when 3.7 dies in 6 months

File: tests/test_modeling.py (hunk 8 above)
Comment (lvwerra) [reply tests/test_modeling.py:94]: I'd address this too in a separate PR.

File: .github/workflows/tests.yml (hunk 5 above)
Comment (lvwerra) [reply .github/workflows/tests.yml:18]: sounds good

File: .github/workflows/tests.yml (hunk 1 above)
Comment (lvwerra) [reply .github/workflows/tests.yml:25]: done

File: tests/test_modeling.py (hunk 8 above)
Comment (younesbelkada) [reply tests/test_modeling.py:94]: Addressed in https://github.com/lvwerra/trl/pull/65 !

File: tests/test_modeling.py (hunk 7 above)
Comment (younesbelkada) [reply tests/test_modeling.py:78]: Addressed in https://github.com/lvwerra/trl/pull/65 !

**Comment by lvwerra** [comment]
All comments should be addressed. Also applied the quality to the recent merges.

**Review by younesbelkada — APPROVED** [review]
Thanks a lot for taking care of the code quality!
